crates/sui-types/src/crypto.rs

```diff
@@ -262,77 +266,207 @@ impl std::fmt::Debug for Signature {
             SuiError::InvalidSignature {
                 error: err.to_string(),
             }
         })?;
 
-        // perform cryptographic signature check
-        public_key
-            .verify(&message, &signature)
-            .map_err(|error| SuiError::InvalidSignature {
-                error: error.to_string(),
+        Ok((signature, pk))
+    }
+
+    fn new(kp: &Self::KeyPair, message: &[u8]) -> SuiResult<Self> {
+        let sig = kp
+            .try_sign(message)
+            .map_err(|_| SuiError::InvalidSignature {
+                error: "Failed to sign valid message with keypair".to_string(),
+            })?;
+
+        let mut signature_bytes: Vec<u8> = Vec::new();
+        signature_bytes.extend_from_slice(&Self::FLAG);
+        signature_bytes.extend_from_slice(sig.as_ref());
+        signature_bytes.extend_from_slice(kp.public().as_ref());
+        Self::from_bytes(&signature_bytes[..]).map_err(|_| SuiError::InvalidSignature {
+            error: format!("Invalid signature"),
+        })
+    }
+}
+
+#[enum_dispatch(Signature)]
+pub trait SuiSignature: Sized + signature::Signature {
+    fn signature_bytes(&self) -> &[u8];
+    fn public_key_bytes(&self) -> &[u8];
+
+    fn flag_bytes(&self) -> &[u8] {
+        &self.as_ref()[..FLAG_LENGTH]
+    }
+
+    fn verify<T>(&self, value: &T, author: SuiAddress) -> SuiResult<()>
+    where
+        T: Signable<Vec<u8>>;
+
+    fn add_to_verification_obligation_or_verify(
+        &self,
+        author: SuiAddress,
+        obligation: &mut VerificationObligation,
+        idx: usize,
+    ) -> SuiResult<()>;
+}
+
+impl<S: SuiSignatureInner + Sized> SuiSignature for S {
+    fn verify<T>(&self, value: &T, author: SuiAddress) -> SuiResult<()>
+    where
+        T: Signable<Vec<u8>>,
+    {
+        // Currently done twice - can we improve on this?;
+        let (sig, pk) = &self.get_verification_inputs(author)?;
+        let mut message = Vec::new();
+        value.write(&mut message);
+        pk.verify(&message[..], sig)
+            .map_err(|_| SuiError::InvalidSignature {
+                error: "hello".to_string(),
             })
     }
 
-    pub fn get_verification_inputs(
+    fn add_to_verification_obligation_or_verify(
         &self,
         author: SuiAddress,
-    ) -> Result<(AccountSignature, PublicKeyBytes), SuiError> {
-        // Is this signature emitted by the expected author?
-        let public_key_bytes: PublicKeyBytes =
-            PublicKeyBytes::from_bytes(self.public_key_bytes()).expect("byte lengths match");
-        let received_addr: SuiAddress = (&public_key_bytes).into();
-        if received_addr != author {
-            return Err(SuiError::IncorrectSigner {
-                error: format!("Signature get_verification_inputs() failure. Author is {author}, received address is {received_addr}")
-            });
+        obligation: &mut VerificationObligation,
+        idx: usize,
+    ) -> SuiResult<()> 
```

Comment:
Call this function to either add to obligation (if batching is possible), or verify individually.